src/tvdb/client.py

```diff
@@ -0,0 +1,192 @@
+from abc import ABC, abstractmethod
+from typing import Any, ClassVar, Literal, overload, override
+
+import aiohttp
+from yarl import URL
+
+from src.settings import TVDB_API_KEY
+from src.tvdb.generated_models import (
+    MovieBaseRecord,
+    MovieExtendedRecord,
+    SearchResult,
+    SeriesBaseRecord,
+    SeriesExtendedRecord,
+)
+from src.tvdb.models import MovieResponse, SearchResponse, SeriesResponse
+from src.utils.log import get_logger
+
+log = get_logger(__name__)
+
+type JSON_DATA = dict[str, JSON_DATA] | list[JSON_DATA] | str | int | float | bool | None  # noice
+
+
+class _Media(ABC):
+    def __init__(self, client: "TvdbClient"):
+        self.client = client
+        self._id: int | None = None
+
+    @property
+    def id(self) -> int | str | None:
+        return self._id
+
+    @id.setter
+    def id(self, value: int | str) -> None:
+        self._id = int(str(value).split("-")[1])
+
+    def __call__(self, media_id: str) -> "_Media":
+        self.id = media_id
+        return self
+
+    @abstractmethod
+    async def search(self, search_query: str, limit: int = 1) -> list[Any]: ...
+
+    @abstractmethod
+    async def fetch(self, *, extended: bool = False) -> Any: ...
+
+
+class Series(_Media):
```

Comment:
This class is kind of weird IMO, it's half representing a concrete series, when using the id and fetch, and half just allowing you to interact with the API by searching for arbitrary series

Seems like this class serves more as a name-space just to hold these functions than it does as an initializable class. Using classes as name-spaces is an anti-pattern, instead, those belong in separate files. With the `fetch` function just taking in that `id` as a param. Alternatively, these could also just be classmethods on the appropriate `models.py` classes, that provide an alternative way to construct that class, like `SearchResult.fetch("xyz", limit=2)` that would return a `list[Self]`.

Same thing about the `Movie` class.

I might just be misunderstanding something here though.

Reply:
I was trying to find a way not to have to pass in the client into the fetch command, this is why the init gives the class the itself as the client, and `__call__` takes the id as param. I was planning on adding the same concept recusrively onto `Series` for `Seasons` and then `Episodes`. I didn't feel like `client.fetch_series` then `client.fetch_series_season` having to pass the id every time would work.

Lmk please if you have a better idea

Reply:
Well, I was thinking of simply doing `Series.fetch(client)`, ``Seasons.fetch(client)`, etc. I think this is a bit cleaner and follows better separation of concerns.

Reply:
The problem being that `SearchResult` is never a valid `SeriesBaseRecord` or `MovieBaseRecord`

Reply:
I'll try something I have an idea